llvm/lib/Transforms/Vectorize/VPlanUtils.cpp

```diff
@@ -376,3 +377,20 @@ bool VPBlockUtils::isLatch(const VPBlockBase *VPB,
   return VPB->getNumSuccessors() == 2 &&
          VPBlockUtils::isHeader(VPB->getSuccessors()[1], VPDT);
 }
+
+std::optional<MemoryLocation>
+vputils::getMemoryLocation(const VPRecipeBase &R) {
+  return TypeSwitch<const VPRecipeBase *, std::optional<MemoryLocation>>(&R)
+      .Case<VPWidenStoreRecipe, VPInterleaveBase, VPReplicateRecipe>(
```

Comment:
Will this remain small / exhaustive list of recipes?

Reply:
Yes, it will only include memory related recipes, so the list here is pretty much complete. Currently there's no convenient way to cast from a recipe to VPIRMetadata unfortunately, hence the type switch. 

Once https://github.com/llvm/llvm-project/pull/166245 the type-switch can be replaced by `isa<>` + `cast<VPIRMetadata>`